cirq/devices/graph_device.py

```diff
@@ -0,0 +1,230 @@
+
+from cirq.devices.hypergraph import UndirectedHypergraph
+
+
+class HashQubit(ops.QubitId):
+    """A qubit identified by a hashable value."""
+
+    def __init__(self, value: Hashable) -> None:
+        if not isinstance(value, Hashable):
+            raise TypeError('not isinstance({}, Hashable)'.format(value))
+        self.value = value
+
+    def _comparison_key(self):
+        return self.value
+
+    def __str__(self):
+        return str(self.value)
+
+    def __repr__(self):
+        return 'cirq.devices.graph_device.HashQubit({})'.format(
+                repr(self.value))
+
+
+class UndirectedGraphDeviceEdge(metaclass=abc.ABCMeta):
+    """An edge of an undirected graph device.
+    """
+
+    @abc.abstractmethod
+    def duration_of(self, operation: ops.Operation) -> value.Duration:
+        pass
+
+
+    @abc.abstractmethod
+    def validate_operation(self, operation: ops.Operation) -> None:
+        pass
+
+
+class FixedDurationUndirectedGraphDeviceEdge(UndirectedGraphDeviceEdge):
+    """An edge of an undirected graph device on which every operation is
+    allowed and has the same duration."""
+
+    def __init__(self, duration: value.Duration):
+        self._duration = duration
+
+
+    def duration_of(self, operation: ops.Operation) -> value.Duration:
+        return self._duration
+
+
+    def validate_operation(self, operation: ops.Operation) -> None:
+        pass
+
+class _UnconstrainedUndirectedGraphDeviceEdge(UndirectedGraphDeviceEdge):
+    """A device edge that allows everything."""
+
+    def duration_of(self, operation: ops.Operation) -> value.Duration:
+        return value.Duration(picos=0)
+
+    def validate_operation(self, operation: ops.Operation) -> None:
+        pass
+
+UnconstrainedUndirectedGraphDeviceEdge = (
+        _UnconstrainedUndirectedGraphDeviceEdge())
+
+
+def is_undirected_device_graph(graph: UndirectedHypergraph) -> bool:
+    if not isinstance(graph, UndirectedHypergraph):
+        return False
+    for _, label in graph.labelled_edges.items():
+        if not (label is None or isinstance(label, UndirectedGraphDeviceEdge)):
+            return False
+    return True
+
+def is_crosstalk_graph(graph: UndirectedHypergraph) -> bool:
+    if not isinstance(graph, UndirectedHypergraph):
+        return False
+    for vertex in graph.vertices:
+        if not isinstance(vertex, frozenset):
+            return False
+    for edge, label in graph.labelled_edges.items():
+        if len(edge) < 2:
+            return False
+        if not ((label is None) or callable(label)):
+            return False
+    return True
+
+
+def raise_crosstalk_error(*ops: ops.Operation):
+    raise ValueError('crosstalk on {}'.format(ops))
+
+
+class UndirectedGraphDevice(devices.Device):
+    """A device whose properties are represented by an edge-labelled graph.
+
+    Each (undirected) edge of the device graph is labelled by an
+    UndirectedGraphDeviceEdge or None. None indicates that any operation is
```

Comment:
Shouldn't the edges represent capabilities, not constraints, since capabilities are often much sparser?

Reply:
There are two types of edges (belonging to the two different graphs). Device edges corresponding to sets of qubits do indeed represent capabilities. Crosstalk edges correspond to sets of _device edges_ and represent constraints on the simultaneous capabilities of those device edges.